packages/salesforcedx-vscode-soql/src/editor/queryRunner.ts

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright (c) 2020, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+import { Connection } from '@salesforce/core';
+import { JsonMap } from '@salesforce/ts-types';
+import { QueryResult } from 'jsforce';
+import * as vscode from 'vscode';
+
+export class QueryRunner {
+  constructor(
+    private connection: Connection,
+    private document?: vscode.TextDocument
+  ) {}
+
+  public async runQuery(queryText: string): Promise<JsonMap[]> {
+    try {
+      const rawQueryData = (await this.connection.query(
+        queryText
+      )) as QueryResult<JsonMap>;
+      const cleanQueryRecords = this.flattenQueryData(rawQueryData);
+
+      return cleanQueryRecords;
+    } catch (error) {
+      // TODO: Doc Review, i18n
+      vscode.window.showErrorMessage(`Could not run the Query ${error}`);
+      throw error;
+    }
+  }
+  /*
+  As query complexity grows
+  we will need to flatten the results of nested values
+  in order to be parsed and diplayed correctly
+  */
+  private flattenQueryData(rawqQueryData: QueryResult<JsonMap>) {
+    const records = rawqQueryData.records;
+    // filter out the attributes key
+    records.forEach(result => delete result.attributes);
+    return records;
```

Comment:
how about  getting rid of an intermediate variable records and just return rawqQueryData.forEach(result => delete result.attributes);

Reply:
I was hesitant to do that, forEach() has got me in the past because it [does not return any value](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach). I think I need to mutate a value in memory and then return that value.

>forEach() executes the callback function once for each array element; unlike map() or reduce() it always returns the value undefined and is not chainable. The typical use case is to execute side effects at the end of a chain.

Reply:
@jonnyhork another way to do it is `return rawQueryData.map(({ attributes, ...props }) => props)` , see [here](https://stackoverflow.com/a/46839399)

Reply:
my bad. i didn't read the code properly.

Reply:
Thanks @xyc, I like the looks of that! Feel like I've been told mutating state is bad juju anyway.